tensorboard/webapp/notification_center/_data_source/notification_center_data_source.ts

```diff
@@ -30,6 +31,24 @@ export class TBNotificationCenterDataSource
   constructor(private readonly http: TBHttpClient) {}
 
   fetchNotifications(): Observable<NotificationCenterResponse> {
-    return this.http.get<NotificationCenterResponse>(`/data/notifications`);
+    return this.http
+      .get<NotificationCenterResponse>(`/data/notifications`)
+      .pipe(
+        map((response) => {
+          const todayInMs: number = new Date(
+            new Date().toDateString()
+          ).getTime();
```

Comment:
can we use `(new Date()).getTime()`?
(which returns the current time)
Or do you feel strongly about 12AM?

Reply:
I think so. Imagine this:
1. there is a notification at 9pm today
2. Now is 1pm so when fetched data, it gets clipped to 1pm.
3. you first load TensorBoard and read notification at 1pm today
4. Wait 2hrs and now reload TensorBoard
5. Now notifications are clipped at 3pm while you last read notification at 1pm so red dot shows up.

Reply:
Doesn't clipping to 12am still result in problems if the notification is more than 1 day in the future?

1. it's Monday at 9am
2. notification scheduled for Wednesday at 9am
3. notification timestamp is clipped to 12am Monday
4. user successfully marks notification as read
5. Tuesday morning: notification timestamp is clipped to 12am Tuesday
6. user sees unread notification again

IMO it seems better if we fix this further downstream (not in the fetching phase) by just not showing notifications at all if they're still in the future.  That would let us use future-timestamped notifications to deliberately roll something out at a precise time if we needed that for some reason.

Or to put it another way, we'd have 3 possible states for notifications, A) in the future, B) in the past but more recent than last read timestamp, or C) in the past and older than last read timestamp.

The unread count / red dot only considers notifications in state B, and the actual notification panel only shows notifications in states B and C, so notifications in state A are entirely hidden.

Reply:
Luckily, we have been good with at least getting the date right so I am not very worried about that. 

>  deliberately roll something out at a precise time if we needed that for some reason.

Rolling something out at certain date makes sense but I am not sure if we want to have a precise control over the millisecond precision here. Moreover, releases often is accompanied by a feature flag flip which is not time controlled so you will need to have a release anyways.